numpy/fft/_pocketfft.py

```diff
@@ -46,33 +46,36 @@
 # divided. This replaces the original, more intuitive 'fct` parameter to avoid
 # divisions by zero (or alternatively additional checks) in the case of
 # zero-length axes during its computation.
-def _raw_fft(a, n, axis, is_real, is_forward, inv_norm):
+def _raw_fft(a, n, axis, is_real, is_forward, inv_norm, out=None):
     axis = normalize_axis_index(axis, a.ndim)
     if n is None:
         n = a.shape[axis]
 
     fct = 1/inv_norm
 
-    if a.shape[axis] != n:
-        s = list(a.shape)
-        index = [slice(None)]*len(s)
-        if s[axis] > n:
-            index[axis] = slice(0, n)
-            a = a[tuple(index)]
+    n_out = n
+    if is_real:
+        if is_forward:
+            ufunc = pfu.rfft_n_even if n % 2 == 0 else pfu.rfft_n_odd
+            n_out = n // 2 + 1
         else:
-            index[axis] = slice(0, s[axis])
-            s[axis] = n
-            z = zeros(s, a.dtype.char)
-            z[tuple(index)] = a
-            a = z
-
-    if axis == a.ndim-1:
-        r = pfi.execute(a, is_real, is_forward, fct)
+            ufunc = pfu.irfft
     else:
-        a = swapaxes(a, axis, -1)
-        r = pfi.execute(a, is_real, is_forward, fct)
-        r = swapaxes(r, axis, -1)
-    return r
+        ufunc = pfu.fft if is_forward else pfu.ifft
+
+    if out is None:
+        out = empty(a.shape[:axis] + (n_out,) + a.shape[axis+1:],
+                    dtype=complex if is_forward or not is_real else float)
+    elif ((shape := getattr(out, "shape", None)) is not None
+          and (len(shape) != a.ndim or shape[axis] != n_out)):
+        raise ValueError("output array has wrong shape.")
+    # Note: for backward compatibility, we want to accept longdouble as well,
+    # even though it is at reduced precision. To tell the promotor that we
+    # want to do that, we set the signature (to the only the ufunc has).
+    # Then, the default casting='same_kind' will take care of the rest.
+    # TODO: create separate float, double, and longdouble loops.
+    return ufunc(a, fct, axes=[(axis,), (), (axis,)], out=out,
+                 signature=ufunc.types[0])
```

Comment:
Here, both input and output can be complex or float, so I'd need to get that out of the `if/else` that decides the `ufunc` above. It seemed easier to just get it from the ufunc, but I admit I was assuming that an upgrade of `pocketfft` would do both `float32` and `longdouble`, so that one could just rely on regular promotion.